knative/describer.go

```diff
@@ -0,0 +1,100 @@
+package knative
+
+import (
+	"fmt"
+	"github.com/boson-project/faas/client"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+	"knative.dev/eventing/pkg/apis/eventing/v1alpha1"
+	eventingv1client "knative.dev/eventing/pkg/client/clientset/versioned/typed/eventing/v1alpha1"
+	servingv1client "knative.dev/serving/pkg/client/clientset/versioned/typed/serving/v1alpha1"
+)
+
+type Describer struct {
+	Verbose        bool
+	namespace      string
+	servingClient  *servingv1client.ServingV1alpha1Client
+	eventingClient *eventingv1client.EventingV1alpha1Client
+	config 		   *rest.Config
+}
+
+func NewDescriber(namespace string) (describer *Describer, err error) {
+	describer = &Describer{namespace: namespace}
+	loadingRules := clientcmd.NewDefaultClientConfigLoadingRules()
+	clientConfig := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(loadingRules, &clientcmd.ConfigOverrides{})
+	if describer.namespace == "" {
+		namespace, _, err := clientConfig.Namespace()
+		if err == nil {
+			describer.namespace = namespace
+		}
+	}
+	config, err := clientConfig.ClientConfig()
+	if err != nil {
+		return
+	}
+	describer.servingClient, err = servingv1client.NewForConfig(config)
+	if err != nil {
+		return
+	}
+	describer.eventingClient, err = eventingv1client.NewForConfig(config)
+	if err != nil {
+		return
+	}
+	describer.config = config
+	return
+}
+
+func (describer *Describer) Describe(name string) (description client.FunctionDescription, err error) {
+
+	namespace := describer.namespace
+	servingClient := describer.servingClient
+	eventingClient := describer.eventingClient
+
+	service, err := servingClient.Services(namespace).Get(name, metav1.GetOptions{})
+	if err != nil {
+		return
+	}
+
+	serviceLabel := fmt.Sprintf("serving.knative.dev/service=%s", name)
+	routes, err := servingClient.Routes(namespace).List(metav1.ListOptions{LabelSelector: serviceLabel})
```

Comment:
I agree with your earlier comment that we probably want to assume a 1-to-1 between function and route at this point; though we may eat our words later.